src/collections/componentSet.ts

```diff
@@ -56,15 +57,26 @@ export class ComponentSet extends LazyCollection<MetadataComponent> {
    */
   public sourceApiVersion: string;
   public fullName?: string;
+  private logger: Logger;
   private registry: RegistryAccess;
   private components = new Map<string, Map<string, SourceComponent>>();
 
+  // internal component maps used by this.getObject() when building manifests.
+  private destructiveComponents = new Map<string, Map<string, SourceComponent>>();
+  private manifestComponents = new Map<string, Map<string, SourceComponent>>();
+
   public constructor(components: Iterable<ComponentLike> = [], registry = new RegistryAccess()) {
     super();
     this.registry = registry;
     this.apiVersion = this.registry.apiVersion;
+    this.logger = Logger.childFromRoot(this.constructor.name);
+
     for (const component of components) {
-      this.add(component);
+      let asDeletion = false;
+      if (component instanceof SourceComponent) {
```

Comment:
Because you can add `ComponentLike` as well as `SourceComponent`.  It doesn't make sense to mark `ComponentLike` as a delete because it means there's no local source file.